pkg/app/ops/insightcollector/deployments.go

```diff
@@ -30,129 +27,131 @@ import (
-				zap.Error(err))
+	for id, ds := range apps {
+		if err := c.updateApplicationChunks(ctx, ds[0].ProjectId, id, ds, model.InsightMetricsKind_CHANGE_FAILURE_RATE, target); err != nil {
+			c.logger.Error("failed to update application chunks", zap.Error(err))
 			updateErr = err
 		}
 	}
-
-	if updateErr != nil {
-		return updateErr
-	}
-
-	m.DeploymentCreatedAtMilestone = targetDate.Unix()
-	return i.insightstore.PutMilestone(ctx, m)
-}
-
-func (i *InsightCollector) ProcessNewlyCompletedDeployments(ctx context.Context) error {
-	now := time.Now()
-	targetDate := time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, time.UTC)
-	m, err := i.insightstore.LoadMilestone(ctx)
-	if err != nil {
-		if !errors.Is(err, filestore.ErrNotFound) {
-			return err
-		}
-		m = &insight.Milestone{}
-	}
-
-	dc, err := i.findDeploymentsCompletedInRange(ctx, m.DeploymentCompletedAtMilestone, targetDate.Unix())
-	if err != nil {
-		return err
-	}
-
-	var updateErr error
-	for _, f := range i.collectFuncs.developmentCollectFns.newlyCompletedDeploymentsFns {
-		if err := f(ctx, dc, targetDate); err != nil {
-			fv := reflect.ValueOf(f)
-			i.logger.Error("failed to execute a collect function",
-				zap.String("function name", runtime.FuncForPC(fv.Pointer()).Name()),
-				zap.Error(err))
+	for id, ds := range projects {
+		if err := c.updateApplicationChunks(ctx, id, ds[0].ApplicationId, ds, model.InsightMetricsKind_CHANGE_FAILURE_RATE, target); err != nil {
+			c.logger.Error("failed to update application chunks", zap.Error(err))
 			updateErr = err
 		}
 	}
 
-	if updateErr != nil {
-		return updateErr
-	}
-
-	m.DeploymentCompletedAtMilestone = targetDate.Unix()
-	return i.insightstore.PutMilestone(ctx, m)
+	return updateErr
 }
 
-func (i *InsightCollector) collectChangeFailureRate(ctx context.Context, ds []*model.Deployment, target time.Time) error {
-	apps, projects := i.groupDeployments(ds)
+func (c *InsightCollector) collectDevelopmentFrequency(ctx context.Context, ds []*model.Deployment, target time.Time) error {
+	apps, projects := groupDeployments(ds)
 
 	var updateErr error
 	for id, ds := range apps {
-		if err := i.updateApplicationChunks(ctx, ds[0].ProjectId, id, ds, model.InsightMetricsKind_CHANGE_FAILURE_RATE, target); err != nil {
-			i.logger.Error("failed to update application chunks", zap.Error(err))
+		if err := c.updateApplicationChunks(ctx, ds[0].ProjectId, id, ds, model.InsightMetricsKind_DEPLOYMENT_FREQUENCY, target); err != nil {
+			c.logger.Error("failed to update application chunks", zap.Error(err))
 			updateErr = err
 		}
 	}
 	for id, ds := range projects {
-		if err := i.updateApplicationChunks(ctx, id, ds[0].ApplicationId, ds, model.InsightMetricsKind_CHANGE_FAILURE_RATE, target); err != nil {
-			i.logger.Error("failed to update application chunks", zap.Error(err))
+		if err := c.updateApplicationChunks(ctx, id, "", ds, model.InsightMetricsKind_DEPLOYMENT_FREQUENCY, target); err != nil {
```

Comment:
After the update of `groupDeployments` logic, I finally got the point here 😂 
Just to confirm, is it okay to just pass an empty string here as `appID` value? 🤔

Reply:
Oh no. Don't know why this happened. Nice catch!

Reply:
looks like it buggy from start (before this refactor) 😂 Let's rethink the behaviors of this `updateApplicationChunks` method 👍

Reply:
Fixed this change.

Let me keep this PR as its title, just doing the refactoring and keep its logic as before.
And I will check the logic later and make another PR if needed.

Reply:
🙆‍♀️ Please make an issue for this later 🙏